src/bun.js/bindings/PCRE2RegExp.cpp

```diff
@@ -0,0 +1,790 @@
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+
+    auto *thisValue = jsDynamicCast<PCRE2RegExp*>(callFrame->thisValue());
+    if (!thisValue)
+        return JSValue::encode(jsUndefined());
+
+    JSValue arg = callFrame->argument(0);
+    if (!arg || arg.isUndefinedOrNull()) {
+        thisValue->m_lastIndex = 0;
+        return JSValue::encode(jsNull());
+    }
+
+    WTF::String string = to16Bit(arg, globalObject, ""_s);
+    RETURN_IF_EXCEPTION(scope, JSValue::encode({}));
+
+    pcre2_match_data_16* matchData = pcre2_match_data_create_from_pattern_16(thisValue->m_regExpCode, NULL);
+
+    int matchResult = pcre2_match_16(
+        thisValue->m_regExpCode,
+        reinterpret_cast<PCRE2_SPTR16>(string.characters16()),
+        string.length(),
+        thisValue->m_lastIndex,
+        0,
+        matchData,
+        NULL
+    );
+
+    if (matchResult < 0) {
+        // catch errors here?
+
+        // switch(matchResult) {
+        //     case PCRE2_ERROR_NOMATCH: {
+        //         return JSValue::encode(jsBoolean(false));
+        //     }
+        //     default: {
+        //         return JSValue::encode(jsBoolean(false));
+        //     }
+        // }
+        return JSValue::encode(jsNull());
+    }
+
+    size_t* outVector = pcre2_get_ovector_pointer_16(matchData);
+
+    if (matchResult == 0) {
+        // no matches
+        pcre2_match_data_free_16(matchData);
+        return JSValue::encode(jsNull());
+    }
+
+    PCRE2_SPTR16 str = reinterpret_cast<PCRE2_SPTR16>(string.characters16());
+
+    size_t count = pcre2_get_ovector_count_16(matchData);
+
+    JSArray* indicesArray = constructEmptyArray(globalObject, nullptr, 0);
+    RETURN_IF_EXCEPTION(scope, JSValue::encode({}));
+    JSArray* result = constructEmptyArray(globalObject, nullptr, 0);
+    RETURN_IF_EXCEPTION(scope, JSValue::encode({}));
+    
+    result->putDirect(vm, vm.propertyNames->index, jsNumber(outVector[0]));
+    result->putDirect(vm, vm.propertyNames->input, jsString(vm, string));
+    result->putDirect(vm, vm.propertyNames->groups, jsUndefined());
+
+    for (size_t i = 0; i < count; i++) {
+        PCRE2_SPTR16 substringStart = str + outVector[2 * i];
+        PCRE2_SIZE substringLength = outVector[2 * i + 1] - outVector[2 * i];
+        UChar *ptr;
+        WTF::String outString;
+        
+        if (substringLength > 0) {
+            outString = WTF::String::createUninitialized(static_cast<unsigned int>(substringLength), ptr); 
+            if (UNLIKELY(!ptr)) {
+                throwOutOfMemoryError(globalObject, scope);
+                pcre2_match_data_free_16(matchData);
+                return JSValue::encode(jsNull());
+            }
+        
+            memcpy(ptr, substringStart, substringLength * sizeof(UChar));;
+        }
+
+        result->putDirectIndex(globalObject, i, jsString(vm, outString));
+
+        JSArray* indices = constructEmptyArray(globalObject, nullptr, 2);
```

Comment:
There's a faster way to construct short arrays like this, but I think it's fine to ignore for now